src/base_alloc/base_alloc_global.c

```diff
@@ -82,41 +83,121 @@ static int size_to_idx(size_t size) {
     return index;
 }
 
-void *umf_ba_global_alloc(size_t size) {
+// stores metadata just before 'ptr' and returns beginning of usable
+// space to the user. metadata consists of 'size' that is the allocation
+// size and 'offset' that specifies how far is the returned ptr from
+// the origin ptr (used for aligned alloc)
+static void *transform_ptr(void *ptr, size_t size, size_t alignment) {
+    assert(size < (1ULL << 32));
+    assert(alignment < (1ULL << 32));
+    assert(ptr);
+
+    void *user_ptr;
+    if (alignment <= sizeof(size_t)) {
+        user_ptr = (void *)((uintptr_t)ptr + sizeof(size_t));
+    } else {
+        user_ptr = (void *)ALIGN_UP((uintptr_t)ptr + sizeof(size_t), alignment);
+    }
+
+    size_t ptr_offset_from_original = (uintptr_t)user_ptr - (uintptr_t)ptr;
+
+    size_t *metadata_loc = (size_t *)((char *)user_ptr - sizeof(size_t));
+    *metadata_loc = size | (ptr_offset_from_original << 32);
```

Comment:
please assert that the bits we truncate are zeros

Reply:
This is already asserted above `assert(size < (1ULL << 32));` and `assert(alignment < (1ULL << 32));`

Reply:
I meant the check for ptr_offset_from_original.

assert(ptr_offset_from_original < (1ULL << 32));

Reply:
Since alignment is checked to be < (1 << 32) the offset can't be bigger than that. But done, I've added the check anyway,